internal/expr/prepare.go

```diff
@@ -270,23 +269,66 @@ func (pe *ParsedExpr) Prepare(args ...any) (expr *PreparedExpr, err error) {
 				}
 				typeMemberPresent[tm] = true
 			}
+			outputs = append(outputs, typeMembers...)
+			outputCols = append(outputCols, outCols)
+		case *bypassPart:
+		default:
+			return nil, fmt.Errorf("internal error: unknown query part type %T", part)
+		}
+	}
+	return &PreparedExpr{inputs: inputs, outputs: outputs, queryParts: pe.queryParts, outputCols: outputCols}, nil
+}
+
+// stmtCriterion contains information that differentiates the SQL strings that
+// can be generated from a single SQLair prepared statement.
+type stmtCriterion struct {
+	sliceLens []int
+}
+
+func (pe *PreparedExpr) sql(sc *stmtCriterion) string {
+	var sql bytes.Buffer
+	var inCount int
+	var outCount int
+	var outputPartCount int
+	var sliceCount int
```

Comment:
I like this implementation but I can be hard to keep track of all the indexes, maybe it makes sense to couple this tighter, for example we can make `outputCols` a slice indexed by a pointer to a queryPart. If we used maps for all of them we can just check based on the *part, which is the thing we are iterating over.

Reply:
I've changed this though I'm in two minds about it. On one hand the map feels a bit like overkill for this, we can just count the position, its very much something we control, so why not just use it? On the other hand its a very easy place to introduce errors, I don't want to optimize something for the sake of it when it solves far more problems to just use the safe, more expensive option.